src/tag.jl

```diff
@@ -0,0 +1,56 @@
+import ForwardDiff: Dual
+
+struct Tag{Parent} end
+
+_find_dual(ctx::Type{T}, i) where {T} = 0
+_find_dual(ctx::Type{T}, i, x::Type{<:Dual{T}}, xs...) where {T} = i
+_find_dual(ctx::Type{T}, i, x, xs...) where {T} = _find_dual(ctx, i-1, xs...)
+
+innertagtype(::Type{Tag{T}}) where T = T
+Base.@pure @inline @generated function find_dual(::Type{T}, xs...) where {T<:Tag}
```

Comment:
Does `@pure` work in `@generated`. What are you trying to make pure? The generator, or the generated body?

Reply:
Neither of those are `@pure`

Splatting is not `@pure` esp with no tight type-contraints

Reply:
I removed the `@generated`s... this function seems to infer OK.

But! `S = tagtype(fieldtype(typeof(args), idx))` creates a `Box` although `idx` is known statically (return value of `find_dual`).

This used to be 115 lines long, got it down to 20. My goal is to compile away any overhead for this function. If you see below there's only 1 `setindex!` call to `box.contents` and it's always executed! So the information does exist...
```julia
julia> using ForwardDiff2: Dual,dualcontext,overdub,Tag

julia> f(x,y)=y
f (generic function with 1 method)

julia> @code_warntype overdub(ctx, f, 1, Dual{Tag{Nothing}}(2, 1))
Body::Dual{Tag{Nothing},Int64,1}
1 ── %1  = %new(Core.Box)::Core.Box
└───       goto #3 if not false
2 ──       nothing
3 ┄─       goto #5 if not false
4 ──       nothing
5 ┄─       (Core.setfield!)(%1, :contents, Tag{Nothing})
│    %7  = %new(ForwardDiff2.:(##16#19), %1)::getfield(ForwardDiff2, Symbol("##16#19"))
│    %8  = (Base.getfield)(args, 1, true)::Int64
│          invoke %7(%8::Int64)
│    %10 = (Base.getfield)(args, 2, true)::Dual{Tag{Nothing},Int64,1}
│    %11 = invoke %7(%10::Dual{Tag{Nothing},Int64,1})::Union{Int64, Dual{Tag{Nothing},Int64,1}}
│    %12 = (isa)(%11, Int64)::Bool
└───       goto #7 if not %12
6 ──       goto #10
7 ── %15 = (isa)(%11, Dual{Tag{Nothing},Int64,1})::Bool
└───       goto #9 if not %15
8 ──       goto #10
9 ──       (Core.throw)(ErrorException("fatal error in type inference (type bound)"))
└───       $(Expr(:unreachable))
10 ┄ %20 = (getfield)(args, 2)::Dual{Tag{Nothing},Int64,1}
└───       return %20

```

ideal code:

```julia
julia> @code_warntype overdub(ctx, f, 1, Dual{Tag{Nothing}}(2, 1))
Body::Dual{Tag{Nothing},Int64,1}
1 ─      goto #3 if not false
2 ─      nothing
3 ┄ %3 = (getfield)(args, 2)::Dual{Tag{Nothing},Int64,1}
└──      return %3

```
> Splatting is not @pure esp with no tight type-contraints

Could you explain? and/or point to a reference if you have?

Reply:
I suspected that the closure in `map(x->_value(Tag{S}, x), args)` maybe the cause of S becoming boxed, but then I tried creating a callable `V{S}()` which behaves like `x->_value(S, x)` and `S` is still boxed:

```julia
julia> @code_warntype overdub(ctx, f, 1, Dual{Tag{Nothing}}(2, 1))
Body::Dual{Tag{Nothing},Int64,1}
1 ── %1  = %new(Core.Box)::Core.Box
└───       goto #3 if not false
2 ──       nothing
3 ┄─       goto #5 if not false
4 ──       nothing
5 ┄─       (Core.setfield!)(%1, :contents, Tag{Nothing})
│    %7  = (Core.isdefined)(%1, :contents)::Bool
└───       goto #7 if not %7
6 ──       goto #8
7 ──       $(Expr(:throw_undef_if_not, :S, false))
8 ┄─ %11 = (Core.getfield)(%1, :contents)::Any
│    %12 = (Core.apply_type)(ForwardDiff2.V, %11)::Type{ForwardDiff2.V{_1}} where _1
│    %13 = (%12)()::ForwardDiff2.V{_1} where _1
│    %14 = (ForwardDiff2.map)(%13, args)::Tuple{Int64,Union{Int64, Dual{Tag{Nothing},Int64,1}}}
│    %15 = (getfield)(%14, 2)::Union{Int64, Dual{Tag{Nothing},Int64,1}}
│    %16 = (isa)(%15, Int64)::Bool
└───       goto #10 if not %16
9 ──       goto #13
10 ─ %19 = (isa)(%15, Dual{Tag{Nothing},Int64,1})::Bool
└───       goto #12 if not %19
11 ─       goto #13
12 ─       (Core.throw)(ErrorException("fatal error in type inference (type bound)"))
└───       $(Expr(:unreachable))
13 ┄ %24 = (getfield)(args, 2)::Dual{Tag{Nothing},Int64,1}
└───       return %24
```

Reply:
It turns out it is because I use `S` inside closures! There's one more closure I missed (multiple outputs case).